compiler/noirc_evaluator/src/ssa/opt/loop_invariant.rs

```diff
@@ -236,34 +253,70 @@ impl<'f> LoopInvariantContext<'f> {
                 }
             }
             Instruction::Binary(binary) => {
-                if !matches!(binary.operator, BinaryOp::Add { .. } | BinaryOp::Mul { .. }) {
-                    return false;
-                }
-
-                let operand_type =
-                    self.inserter.function.dfg.type_of_value(binary.lhs).unwrap_numeric();
-
-                let lhs_const =
-                    self.inserter.function.dfg.get_numeric_constant_with_type(binary.lhs);
-                let rhs_const =
-                    self.inserter.function.dfg.get_numeric_constant_with_type(binary.rhs);
-                let (lhs, rhs) = match (
-                    lhs_const,
-                    rhs_const,
-                    self.outer_induction_variables.get(&binary.lhs),
-                    self.outer_induction_variables.get(&binary.rhs),
-                ) {
-                    (Some((lhs, _)), None, None, Some(upper_bound)) => (lhs, *upper_bound),
-                    (None, Some((rhs, _)), Some(upper_bound), None) => (*upper_bound, rhs),
-                    _ => return false,
-                };
-
-                eval_constant_binary_op(lhs, rhs, binary.operator, operand_type).is_some()
+                self.can_evaluate_binary_op(binary, &self.outer_induction_variables)
             }
             _ => false,
         }
     }
 
+    /// Binary operations can take advantage of that our induction variable has a fixed maximum,
+    /// to be transformed from a checked operation to an unchecked operation.
+    ///
+    /// Checked operations require more bytecode and thus we aim to minimize their usage wherever possible.
+    ///
+    /// If one side of a binary operation is a constant and the other is an induction variable
+    /// with a known upper bound, we know whether that binary operation will ever overflow.
+    /// If we determine that an overflow is not possible we can convert the checked operation to unchecked.
+    fn transform_to_unchecked_from_upper_bound(&mut self, instruction_id: InstructionId) {
+        let Instruction::Binary(binary) = &self.inserter.function.dfg[instruction_id] else {
+            return;
+        };
+
+        if binary.operator.is_unchecked()
+            || !self.can_evaluate_binary_op(binary, &self.current_induction_variables)
+        {
+            return;
+        }
+
+        if let Instruction::Binary(binary) = &mut self.inserter.function.dfg[instruction_id] {
+            binary.operator = binary.operator.into_unchecked();
+        };
```

Comment:
```suggestion
        binary.operator = binary.operator.into_unchecked();
```

This check looks duplicated.

Reply:
Ah no, you have different mutabilities. Can we not make the first `binary` a mutable reference?

Reply:
`self.can_evaluate_binary_op` borrows immutably so we cannot.